crates/ironrdp-rdpdr/src/pdu/efs.rs

```diff
@@ -0,0 +1,229 @@
+//! PDUs for [[MS-RDPEFS]: Remote Desktop Protocol: File System Virtual Channel Extension](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/34d9de58-b2b5-40b6-b970-f82d4603bdb5)
+
+use std::fmt::Debug;
+
+use ironrdp_pdu::utils::{encoded_str_len, write_string_to_cursor, CharacterSet};
+use ironrdp_pdu::{
+    cursor::{ReadCursor, WriteCursor},
+    PduEncode, PduResult,
+};
+use ironrdp_pdu::{invalid_message_err, unexpected_message_type_err, PduError};
+
+/// [2.2.1.1 Shared Header (RDPDR_HEADER)](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-rdpefs/29d4108f-8163-4a67-8271-e48c4b9c2a7c)
+/// A header that is shared by all RDPDR PDUs.
+#[derive(Debug)]
+pub struct SharedHeader {
+    pub component: Component,
+    pub packet_id: PacketId,
+}
+
+impl SharedHeader {
+    pub fn decode(src: &mut ReadCursor) -> PduResult<Self> {
+        Ok(Self {
+            component: src.read_u16().try_into()?,
```

Comment:
`ensure_size` is missing for decode and encode, this should be done to prevent panics on insufficient input/output buffer

Reply:
2c8308b48da6d8c7e79c41b6e5f28611c97a1faf

Makes sense, but was tedious and error prone. It makes me wonder how difficult might it be to change the signature in `PduEncode` from `fn encode(&self, dst: &mut WriteCursor<'_>) -> PduResult<()>;` to `fn encode(&self, dst: &mut WriteBuf) -> PduResult<()>;` and thus get rid of this (cc @CBenoit)

Reply:
For decoding I don’t think we can avoid checking for bounds. As for encoding, this change would not be too difficult, but I don’t really want that. This would not be inline encoding anymore. I know this property is adding a few difficulties like this, but it’s actually useful. Some of my colleagues want C# for IronRDP, and the current design make it trivial to read from and write into buffers allocated on C# side via FFI calls. `WriteBuf` requires ownership of the buffer.